cli/azd/pkg/auth/manager.go

```diff
@@ -0,0 +1,433 @@
+	}, nil
+}
+
+var ErrNoCurrentUser = errors.New("not logged in, run `azd login` to login")
+
+// EnsureLoggedInCredential uses the credential's GetToken method to ensure an access token can be fetched. If this fails,
+// nil, ErrNoCurrentUser is returned. On success, the token we fetched is returned.
+func EnsureLoggedInCredential(ctx context.Context, credential azcore.TokenCredential) (*azcore.AccessToken, error) {
+	token, err := credential.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: cLoginScopes,
+	})
+	if err != nil {
+		return &azcore.AccessToken{}, ErrNoCurrentUser
+	}
+
+	return &token, nil
+}
+
+// CredentialForCurrentUser returns a TokenCredential instance for the current user. If `auth.useLegacyAzCliAuth` is set to
+// a truthy value in config, an instance of azidentity.AzureCLICredential is returned instead.
+func (m *Manager) CredentialForCurrentUser(ctx context.Context) (azcore.TokenCredential, error) {
+	cfg, err := config.GetUserConfig(m.configManager)
+	if err != nil {
+		return nil, fmt.Errorf("fetching current user: %w", err)
+	}
+
+	if useLegacyAuth, has := cfg.Get(cUseLegacyAzCliAuthKey); has {
+		if use, err := strconv.ParseBool(useLegacyAuth.(string)); err == nil && use {
+			log.Printf("delegating auth to az since %s is set to true", cUseLegacyAzCliAuthKey)
+			cred, err := azidentity.NewAzureCLICredential(nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create credential: %v: %w", err, ErrNoCurrentUser)
+			}
+			return cred, nil
+		}
+	}
+
+	currentUser, err := readUserProperties(cfg)
+	if err != nil {
+		return nil, ErrNoCurrentUser
+	}
+
+	if currentUser.HomeAccountID != nil {
+		for _, account := range m.publicClient.Accounts() {
+			if account.HomeAccountID == *currentUser.HomeAccountID {
+				return newAzdCredential(m.publicClient, &account), nil
+			}
+		}
+	} else if currentUser.TenantID != nil && currentUser.ClientID != nil {
+		ps, err := m.loadSecret(*currentUser.TenantID, *currentUser.ClientID)
+		if err != nil {
+			return nil, fmt.Errorf("loading secret: %v: %w", err, ErrNoCurrentUser)
+		}
+
+		if ps.ClientSecret != nil {
+			cred, err := azidentity.NewClientSecretCredential(
+				*currentUser.TenantID, *currentUser.ClientID, *ps.ClientSecret, nil,
+			)
+
+			if err != nil {
+				return nil, fmt.Errorf("creating credential: %v: %w", err, ErrNoCurrentUser)
+			}
+
+			return cred, nil
+
+		} else if ps.ClientCertificate != nil {
+			certData, err := base64.StdEncoding.DecodeString(*ps.ClientCertificate)
+			if err != nil {
+				return nil, fmt.Errorf("decoding certificate: %w", err)
+			}
+
+			certs, key, err := azidentity.ParseCertificates(certData, nil)
+			if err != nil {
+				return nil, fmt.Errorf("parsing certificate: %w", err)
+			}
+
+			cred, err := azidentity.NewClientCertificateCredential(
+				*currentUser.TenantID, *currentUser.ClientID, certs, key, nil,
+			)
+
+			if err != nil {
+				return nil, fmt.Errorf("creating credential: %v: %w", err, ErrNoCurrentUser)
+			}
```

Comment:
Same thing here -- is it possible that this is a signal that the credential has expired? The error message "user is not logged in" might be confusing to the user.

Reply:
Yes - looking at the error returns from NewClientCertificateCredential, it looks like the places you could get into trouble are if you PEM certificate didn't contain a certificate or had a private key that was not an RSA private key.

In practice we shouldn't hit this case because we only store certificate data once we had successfully created the credential object in `LoginWithServicePrincipalCertificate`.

Reply:
I see. I'm guessing we should provide user-friendly validation errors above then we're checking for valid private-key certificate.